Documentation for the changes in private registry credentials (#509)
okteto/docs: src/content/self-hosted/install/crds.mdx

---
title: Custom Resource Definitions
description: Custom resource definitions installed by Okteto
sidebar_label: Custom Resource Definitions
id: custom-resource-definitions
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

As part of Okteto's installation, some Custom Resource Definitions (CRDs) are installed in your cluster. These CRDs are used by Okteto to mainly store different information. You can see the list of CRDs installed by Okteto exploring the folder `crds` in the templates of the [Okteto Helm Chart](https://artifacthub.io/packages/helm/okteto/okteto) or by running the following command in your cluster:

```console
kubectl get crd | grep okteto.com
```

> You can let Okteto install these CRDs or you can install them manually on your own, but by aware they are needed for Okteto to work properly. You can use the [following configuration setting](/self-hosted/administration/configuration.mdx/#crds) to let Okteto know if it should install them or not. By default, Okteto will install them.

Reviewer: ```suggestion
You can let Okteto install these CRDs or you can install them manually on your own, but be aware they are needed for Okteto to work properly. You can use the [following configuration setting](/self-hosted/administration/configuration.mdx/#crds) to let Okteto know if it should install them or not. By default, Okteto will install them.
```

## Private Registries

> As part of the changes for `1.14`, Okteto has changed the way it handles the configuration for private registries. From that version, they can be configured via the [Admin dashboard](administration/dashboard.mdx/#registry-credentials), or they can be defined directly in your cluster through one of the CRDs installed by Okteto: `admin.okteto.com`. This last case is useful, for example, if you have your own mechanism to provision credentials/secrets in your cluster.

Reviewer: ```suggestion
In Okteto version `1.14` we've changed the way private registry configuration is managed. In `1.14` onward you must configure private registries via the [Admin dashboard](administration/dashboard.mdx/#registry-credentials) or by defining them directly in your cluster through one of the CRDs installed by Okteto: `admin.okteto.com`. This last case is useful, for example, if you have your own mechanism to provision credentials/secrets in your cluster.
```

> If you prefer to define the registries directly in your cluster, you can do it creating an [Opaque Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#creating-a-secret) in the cluster with the data needed for the registry. Depending on the type of registry, you will need different fields. Once the secret is created, you will need to create a resource with the apiVersion `admin.okteto.com/v1`, which is one of the CRDs installed by Okteto, refering to the secret which contains the credentials.

Reviewer: ```suggestion
If you prefer to define the registries directly in your cluster, you can do it by creating an [Opaque Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#creating-a-secret) in the cluster with the data needed for the registry. Depending on the type of registry, you will need different fields. 

Once the secret is created, you will need to create a resource with the apiVersion `admin.okteto.com/v1`, which is one of the CRDs installed by Okteto, referring to the secret which contains the credentials.
```

    userSecretRef:
      key: <field in referenced secret> # key of the secret which contains the username
      name: <secret-name> # name of the secret where the username is stored
```

In the case of AWS, where you have to specify the Access Key and Access Secret, the process is similar to the static credentials one. You will need to create a secret with those values, and then, create a CRD like this (the name of the resource has to be the hostname of the registry to be configured and it has to end with `.amazonaws.com`):

```yaml
apiVersion: admin.okteto.com/v1
kind: Registry
metadata:
  name: 111122223333.dkr.ecr.eu-central-1.amazonaws.com # hostname of the registry to be configured
  namespace: okteto
spec:
  awsIamUser:
    accessKeyIDSecretRef:
      key: <field in referenced secret> # key of the secret which contains the access key
      name: <secret-name> # name of the secret where the access key is stored
    secretAccessKeySecretRef:
      key: <field in referenced secret> # key of the secret which contains the access secret key
      name: <secret-name> # name of the secret where the access secret is stored
```

> > Be aware that any change you do in those credentials might take a few minutes to be propagated to all the components in the cluster. To make sure Okteto is reading properly those registries, you can check it going to the [Admin dashboard](administration/dashboard.mdx/#registry-credentials). They will be displayed, but they can't be modified from the UI. If you want to manage them from the UI, they have to be created from there.

Reviewer: ```suggestion
:::note
Any changes you make to credentials can take a few minutes to propagate to all components in the cluster. To make sure Okteto is able to access the registries, you can check in the [Admin dashboard](administration/dashboard.mdx/#registry-credentials). They will be displayed, but they can't be modified from the UI. If you want to manage them from the UI, they have to be created from there.
:::
```

Is the last sentence still valid with these latest changes? I assume it may be accurate and specific to AWS but wanted to double check.

I may have further suggestions on this copy depending on the clarification here 😄

Author: This note is for both types (static and AWS) and it is still valid. If you create the credentials in this way (using CRDs), you will only see them in the UI but you cannot modify them. If you want to be able to edit them from the UI, you would need to create them from there

Reviewer: Thanks for the clarification 🙇 

I updated my proposed change based on that but let me know if you disagree with it

Author: Your proposal looks good, I just commit it
